ksqldb-engine/src/main/java/io/confluent/ksql/query/QueryRegistryImpl.java

```diff
@@ -176,12 +175,59 @@ public TransientQueryMetadata createTransientQuery(
         excludeTombstones,
         new ListenerImpl(),
         new StreamsBuilder(),
-        endOffsets
+        Optional.empty()
     );
+    query.initialize();
     registerTransientQuery(serviceContext, metaStore, query);
     return query;
   }
 
+  // CHECKSTYLE_RULES.OFF: ParameterNumberCheck
+  @Override
+  public TransientQueryMetadata createStreamPullQuery(
+      final SessionConfig config,
+      final ServiceContext serviceContext,
+      final ProcessingLogContext processingLogContext,
+      final MetaStore metaStore,
+      final String statementText,
+      final QueryId queryId,
+      final Set<SourceName> sources,
+      final ExecutionStep<?> physicalPlan,
+      final String planSummary,
+      final LogicalSchema schema,
+      final OptionalInt limit,
+      final Optional<WindowInfo> windowInfo,
+      final boolean excludeTombstones,
+      final ImmutableMap<TopicPartition, Long> endOffsets) {
+    // CHECKSTYLE_RULES.ON: ParameterNumberCheck
+    final QueryBuilder queryBuilder = queryBuilderFactory.create(
+          config,
+          processingLogContext,
+          serviceContext,
+          metaStore,
+          streams,
+          !sandbox);
+
+    final TransientQueryMetadata query = queryBuilder.buildTransientQuery(
+        statementText,
+        queryId,
+        sources,
+        physicalPlan,
+        planSummary,
+        schema,
+        limit,
+        windowInfo,
+        excludeTombstones,
+        new ListenerImpl(),
+        new StreamsBuilder(),
+        Optional.of(endOffsets)
+    );
+    query.initialize();
+    // We don't register it as a transient query, so it won't show up in `show queries;`,
+    // nor will it count against the push query limit.
```

Comment:
Compare to L182 when we create a transient query, in which we do register it.

This omission right here is why we pushed the stream-pull/transient query split farther down. Since we did that, we were also able to drop a couple of levels of optional end offsets, since stream pull queries always specify them and transient queries never do.